### PR TITLE
Recursively extract tar files (#3)

### PR DESCRIPTION
**Pull request by dsotirho-ucsc**

### CODE REVIEW

File: extract.py [hunk 1]
```diff
@@ -1,31 +1,72 @@
...
+def file_name_minus_extensions(file_name: str) -> Union[str, None]:
+    """
+    Return the file name minus the extension(s)
 
+    >>> file_name_minus_extensions('foo.tar')
+    foo
+    >>> file_name_minus_extensions('foo.tar.gz')
+    foo
+    """
+    if '.' in file_name:
```
Comment (hannes-ucsc) [inline extract.py]: Let's not do two scans here. This can be written just as concise with just .index() and a variable.

File: extract.py [hunk 2]
```diff
@@ -1,31 +1,72 @@
...
+    if tar_path.is_dir():
+        # Run this function recursively for each item found in the given dir
+        logging.debug('Decending into directory %s', tar_path)
+        for file in tar_path.iterdir():
+            extract_tar_file(file)
+    elif tar_path.is_file() and tar_path.name.endswith(('.tar', '.tar.bz', 'tar.gz')):
+        # Extract the tar file and place a completion file in the new folder when done
+        logging.debug('Extracting file %s', tar_path)
+        file_name_stripped = file_name_minus_extensions(tar_path.name)
+        if file_name_stripped and len(file_name_stripped) < len(tar_path.name):
```
Comment (hannes-ucsc) [inline extract.py]: Conflates `''` and `None`.

I don't see how the second term can ever be False. Given the list of extensions above, the inability to strip the extension should constitute an error.

File: extract.py [hunk 3]
```diff
@@ -1,31 +1,72 @@
...
+
+    :param tar_path: Path to a tar file or folder containing one or more tar files
+    """
+    logging.debug('Running extract_tar_file(%s)', tar_path)
+    if tar_path.is_dir():
+        # Run this function recursively for each item found in the given dir
+        logging.debug('Decending into directory %s', tar_path)
+        for file in tar_path.iterdir():
+            extract_tar_file(file)
+    elif tar_path.is_file() and tar_path.name.endswith(('.tar', '.tar.bz', 'tar.gz')):
```
Comment (hannes-ucsc) [inline extract.py]: If the file name ends in those extensions, the fact that the file is not a regular file constitutes an error.

File: extract.py [hunk 4]
```diff
@@ -1,31 +1,72 @@
...
 
 
 def main():
+    # Find the RAW tar files and recursively extract them
     projects_dir = Path.cwd() / 'projects'
     for project_dir in projects_dir.iterdir():
         if project_dir.is_dir() and not project_dir.is_symlink():
             geo_dir = project_dir / 'geo'
             for geo_file in geo_dir.iterdir():
                 if geo_file.name.endswith('_RAW.tar'):
```
Comment (hannes-ucsc) [inline extract.py]: Didn't we say we want to recursively extract all tarballs?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Please stick with the agreed upon approach of two functions. Put the bare extraction function first so the diff looks smaller. Please demo this to me before the requesting another review.

File: extract.py (hunk 3 above)
Comment (hannes-ucsc) [reply extract.py]: Are there any .tar.bz files. If not, drop it. Conventionally, these files are named `.tar.bz2`

Comment (hannes-ucsc) [reply extract.py]: Last but not least, have you thought about the actual extraction of compressed tarballs? I don't think `TarFile` does this for you.

Comment (dsotirho-ucsc) [reply extract.py]: `TarFile` does have some tar file type detection, but I'll specify gzip mode when we have a `.tar.gz` file

File: extract.py (hunk 4 above)
Comment (dsotirho-ucsc) [reply extract.py]: Ah ok, I thought we only wanted to start with the '_RAW' tar files then recursively act on the extracted files.

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**